### PR TITLE
 fix(ci): running out of ports due to port range fragmentation 

### PR DESCRIPTION
**Pull request by dpc**

### CODE REVIEW

**Comment by dpc** [comment]
Wait, this version is stupid.

File: utils/portalloc/src/data/dto.rs [hunk 1]
```diff
@@ -10,9 +10,15 @@ pub struct RangeData {
...
 
+fn default_next() -> u16 {
+    crate::LOW
+}
+
 #[derive(Serialize, Deserialize, Debug, Clone, Default)]
 #[serde(rename_all = "kebab-case")]
 pub struct RootData {
+    #[serde(default = "default_next")]
+    pub next: u16,
```
Comment (dpc) [inline utils/portalloc/src/data/dto.rs:21]: This is the core of the idea - store the pointer around the port range in the state file, to move linearly around it and use the available port range quite optimally.

File: utils/portalloc/src/data/dto.rs [hunk 2]
```diff
@@ -24,26 +30,54 @@ impl RootData {
...
+    ///
+    /// If it does return next address after the range that conflicted.
+    pub fn contains(&self, range: std::ops::Range<u16>) -> Option<u16> {
+        self.keys.range(..range.end).next_back().and_then(|(k, v)| {
             let start = *k;
             let end = start + v.size;
 
-            start < range.end && range.start < end
+            if start < range.end && range.start < end {
+                Some(k + v.size)
```
Comment (bradleystachurski) [inline utils/portalloc/src/data/dto.rs:46]: [nit] This might read more clearly as `Some(end)`, since it's already defined.

**Review by bradleystachurski — APPROVED** [review]

**Comment by bradleystachurski** [comment]
Hmm, the CI logs aren't too helpful but seems to repro locally

```
00:00:06 2024-05-13T06:13:06.064990Z  INFO fm::devimint: Devimint data dir path=/tmp/nix-shell.keInI5/bckn_bitcoind_dummy-Tfi4/devimint-188732-899
00:01:06 Error: Polling btcoind rpc ready failed after 111 retries (timeout: 60s)
00:01:06
00:01:06 Caused by:
00:01:06     JSON-RPC error: transport error: Couldn't connect to host: Connection refused (os error 111)
```

**Review by bradleystachurski — APPROVED** [review]

**Pull request merged**